flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/StateChangeFsUploader.java

```diff
@@ -19,50 +19,30 @@
 
 import org.apache.flink.annotation.VisibleForTesting;
 import org.apache.flink.changelog.fs.StateChangeUploadScheduler.UploadTask;
-import org.apache.flink.core.fs.FSDataOutputStream;
 import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.FileSystem.WriteMode;
 import org.apache.flink.core.fs.Path;
-import org.apache.flink.runtime.state.SnappyStreamCompressionDecorator;
-import org.apache.flink.runtime.state.StreamCompressionDecorator;
 import org.apache.flink.runtime.state.StreamStateHandle;
-import org.apache.flink.runtime.state.UncompressedStreamCompressionDecorator;
 import org.apache.flink.runtime.state.filesystem.FileStateHandle;
-import org.apache.flink.util.clock.Clock;
-import org.apache.flink.util.clock.SystemClock;
 
 import org.apache.flink.shaded.guava30.com.google.common.io.Closer;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.io.BufferedOutputStream;
 import java.io.IOException;
-import java.io.OutputStream;
 import java.util.Collection;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.UUID;
 import java.util.function.BiFunction;
-import java.util.stream.Collectors;
-
-import static org.apache.flink.core.fs.FileSystem.WriteMode.NO_OVERWRITE;
 
 /**
  * A synchronous {@link StateChangeUploadScheduler} implementation that uploads the changes using
  * {@link FileSystem}.
  */
-public class StateChangeFsUploader implements StateChangeUploader {
+public class StateChangeFsUploader extends AbstractStateChangeFsUploader {
```

Comment:
`AbstractStateChangeFsUploader` was extracted and is extended by `DuplicatingStateChangeFsUploader` and `StateChangeFsUploader`.
It has some `protected` methods that are used by descendants; and no `abstract` methods. And the calling methods in descendants (`upload`) have some duplication.

That means there is something wrong with the design.
How about applying template method pattern and having `upload()` in the base class and some `abstract` methods to customize it (i.e. create stream and handle errors)?

(alternatively, the two classes can be just implementers; code can be shared via composition or just static methods)

WDYT?

Reply:
Nice suggestion, I extracted the logic of creating stream into `prepareStream()`.